plugins/woocommerce/tests/e2e-pw/tests/smoke-tests/upload-plugin.spec.js

```diff
@@ -0,0 +1,104 @@
+const {
+	ADMINSTATE,
+	ADMIN_USER,
+	ADMIN_PASSWORD,
+	GITHUB_TOKEN,
+	PLUGIN_NAME,
+	PLUGIN_REPOSITORY,
+} = process.env;
+const { test, expect } = require( '@playwright/test' );
+const path = require( 'path' );
+const {
+	createPlugin,
+	deletePlugin,
+	downloadZip,
+	deleteZip,
+	getLatestReleaseZipUrl,
+} = require( '../../utils/plugin-utils' );
+
+const adminUsername = ADMIN_USER ?? 'admin';
+const adminPassword = ADMIN_PASSWORD ?? 'password';
+
+let pluginZipPath;
+let pluginSlug;
+
+test.describe( `${ PLUGIN_NAME } plugin can be uploaded and activated`, () => {
+	// Skip test if PLUGIN_REPOSITORY is falsy.
+	test.skip(
+		! PLUGIN_REPOSITORY,
+		`Skipping this test because value of PLUGIN_REPOSITORY was falsy: ${ PLUGIN_REPOSITORY }`
+	);
+
+	test.use( { storageState: ADMINSTATE } );
+
+	test.beforeAll( async () => {
+		pluginSlug = PLUGIN_REPOSITORY.split( '/' ).pop();
+
+		// Get the download URL and filename of the plugin
+		const pluginDownloadURL = await getLatestReleaseZipUrl( {
+			repository: PLUGIN_REPOSITORY,
+			authorizationToken: GITHUB_TOKEN,
+		} );
+		const zipFilename = pluginDownloadURL.split( '/' ).pop();
+		pluginZipPath = path.resolve( __dirname, `../../tmp/${ zipFilename }` );
+
+		// Download the needed plugin.
+		await downloadZip( {
+			url: pluginDownloadURL,
+			downloadPath: pluginZipPath,
+			authToken: GITHUB_TOKEN,
+		} );
+	} );
+
+	test.afterAll( async ( { baseURL, playwright } ) => {
+		// Delete the downloaded zip.
+		await deleteZip( pluginZipPath );
+
+		// Delete the plugin from the test site.
+		await deletePlugin( {
+			request: playwright.request,
+			baseURL,
+			slug: pluginSlug,
+			username: adminUsername,
+			password: adminPassword,
+		} );
+	} );
+
+	test( `can upload and activate ${ PLUGIN_NAME }`, async ( {
+		page,
+		playwright,
+		baseURL,
+	} ) => {
+		// Delete the plugin if it's installed.
+		await deletePlugin( {
+			request: playwright.request,
+			baseURL,
+			slug: pluginSlug,
+			username: adminUsername,
+			password: adminPassword,
+		} );
+
+		// Install and activate plugin
+		await createPlugin( {
+			request: playwright.request,
+			baseURL,
+			slug: pluginSlug.split( '/' ).pop(),
+			username: adminUsername,
+			password: adminPassword,
+		} );
```

Comment:
In the Puppeteer tests, the installing and activating of the plugin is done through the UI. But since our "Smoke tests with {PLUGIN} plugin installed" will be using a wpenv environment, uploading of a plugin through the UI will give us the error "The uploaded file exceeds the upload_max_filesize directive in php.ini". So for the meantime, I used the WordPress API in uploading the plugin in wpenv.

Reply:
Yeah, I spent the better part of a day trying to figure out how to fix/work around this -- and I came up dry.  WordPress API works well (although in my test run, I did have a couple of the tests error out).

The only other idea I had was to maybe install the plugins using WP CLI somehow... but that's just a wild thought at the moment.